### PR TITLE
stock configs in /usr/lib/netdata

### LINKED ISSUE #3988: Netdata looks in ../../../../ to get it's config
On Debian Netdata looks at /usr/lib/x86_64-linux-gnu/netdata/plugins.d/../../../../etc/netdata/python.d/postgres.conf to get the config. As you can see this is one ../ too short and it ends up in /usr.

Since the idea is to find the config in /etc/netdata anyway why not use the direct path of /etc/netdata?

I've create PR #3987 for this.

### LINKED ISSUE #4182: stock config files should be in `/usr/lib/netdata/`
Today netdata installs the default config files in `/etc/netdata`. 

This has the problem that the default ones supplied by us and the user edited ones are merged, in one directory structure.

To solve this problem during updates, the netdata installer maintains an MD5 hash of all versions of all config files ever supplied by us, and compares all files in `/etc/netdata` with these hashes to check if any file has been modified.

A better solution would be to install all default config files in `/usr/lib/netdata`. The installer will overwrite them, without checking anything. Any user willing to change a default file, should copy it under `/etc/netdata` and edit it.

The full list of actions is at #4283

### PR DESCRIPTION
**Pull request by ktsaou**
fixes #4182

This PR does the following:

- [x] Makefiles now install stock config files in `/usr/lib/netdata/conf.d`. The structure within this directory is the same with `/etc/netdata`.

- [x] netdata supports configuring this directory from `netdata.conf` via `[global].stock config directory`.

- [x] netdata exports the variable `NETDATA_STOCK_CONFIG_DIR` to let the plugins know.

- [x] netdata exports the variable `NETDATA_USER_CONFIG_DIR`

   the previous variable `NETDATA_CONFIG_DIR` is still supported, but should be deprecated as it is now misleading.

- [x] `netdata.conf`

   first try the filename given with `-c` command line option. If not given (or not found) use `/etc/netdata/netdata.conf`, if not found then `/usr/lib/netdata/conf.d/netdata.conf`

- [x] `stream.conf`

   first try `/etc/netdata/stream.conf`, if not found then `/usr/lib/netdata/conf.d/stream.conf`

- [x] `health.d` (alarms configuration)

   load all *.conf files from `/etc/netdata/health.d` then all *.conf files from `/usr/lib/netdata/conf.d/health.d` that do not exist in the same subpath in `/etc/netdata/health.d`

- [x] `statsd.d` (statsd synthetic charts)

   load all *.conf files from `/etc/netdata/statsd.d` then all *.conf files from `/usr/lib/netdata/conf.d/statsd.d` that do not exist in the same subpath in `/etc/netdata/statsd.d`

   statsd supports including files. Files to included with relative filenames use the same path the file that included them resides. This means that if `/etc/netdata/statsd.d/myapp.conf` includes file `test`, then it will be included from `/etc/netdata/statsd.d/test` only.

- [x] `node.d.plugin`

   first try `/etc/netdata` and if that fails then `/usr/lib/netdata/conf.d`. This happens for both the plugin configuration `node.d.conf` and its modules (under `node.d` directory).

- [x] `python.d.plugin` by @l2isbad 

   It first tries to load config files from `/etc/netdata` and falls back to `/usr/lib/netdata/conf.d`.

- [x] `charts.d.plugin`

   since this a BASH plugin, it first sources the config files found at `/usr/lib/netdata/conf.d` and then it sources the config files found at `/etc/netdata`. So the later overwrite the values of the first.

- [x] `apps.plugin`

   first try `/etc/netdata/apps_groups.conf`, if not found then `/usr/lib/netdata/conf.d/apps_groups.conf`

- [x] `fping.plugin`

   since this a BASH plugin, it first sources the config files found at `/usr/lib/netdata/conf.d` and then it sources the config files found at `/etc/netdata`. So the later overwrite the values of the first.

- [x] `tc-qos-helper.sh`

   since this a BASH plugin, it first sources the config files found at `/usr/lib/netdata/conf.d` and then it sources the config files found at `/etc/netdata`. So the later overwrite the values of the first.

- [x] `cgroup-name.sh`

   custom cgroup names can now be defined in both `/etc/netdata/cgroup-names.conf` and `/var/lib/netdata/conf.d/cgroup-names.conf`, in this order.

- [x] `alarm-notify.sh`

   since this a BASH plugin, it first sources the config files found at `/usr/lib/netdata/conf.d` and then it sources the config files found at `/etc/netdata`. So the later overwrite the values of the first.

- [x] compile time decision of netdata paths, for all plugins (shell, node, python).

   This fixes #3988 (when running plugins by hand, they may fail to find netdata directories).

- [x] adapt `netdata-installer.sh` to remove stock configs from `/etc/netdata`

   before this PR, the installer was checking the MD5 hashes of files in `/etc/netdata` and for those found altered from the stock versions, it was keeping a backup. Then it was installing netdata (`make install`) which overwrote the files, and then it restored the backup it had kept.

   The new logic is much cleaner: the installer deletes from `/etc/netdata` all files that match the stock config files. It checks both the filename and the MD5 hashes, so this should be pretty safe to run. It does this **only once** (so it remembers it has done it in the past and does not do it again).

   After updating netdata with this PR, `/etc/netdata` should contain only the files that do not match stock versions. The link `000.-.USE.THE.orig.LINK.TO.COPY.AND.EDIT.STOCK.CONFIG.FILES` is created only when at least one stock file found in `/etc/netdata` and was deleted. On vanilla installations it won't be there. The `orig` link however, is created every time.

```sh
# ls -l /etc/netdata/
total 1792
lrwxrwxrwx 1 root netdata      23 Sep 26 17:45 000.-.USE.THE.orig.LINK.TO.COPY.AND.EDIT.STOCK.CONFIG.FILES -> /usr/lib/netdata/conf.d
drwxr-xr-x 1 root netdata      56 Sep 26 17:54 charts.d
drwxr-xr-x 1 root netdata       0 Sep 26 17:54 health.d
-rw-r----- 1 root netdata 1782100 Jul  6 20:45 netdata.conf
drwxr-xr-x 1 root netdata     252 Sep 26 17:54 node.d
lrwxrwxrwx 1 root netdata      23 Sep 26 17:45 orig -> /usr/lib/netdata/conf.d
drwxr-xr-x 1 root netdata     190 Sep 26 17:55 python.d
drwxr-xr-x 1 root netdata       0 Sep 26 17:56 statsd.d
```

- [x] adapt `makeself` binary packages

   binary packages were shipped with `etc.new` directory and the static installer was comparing the files already found in `etc/netdata` against the MD5 hashes, to decide if it should overwrite them with the ones in `etc.new/netdata`. This logic is now removed, since the default `etc/netdata` directory in the binary archive is empty (all config files are in `usr/lib/netdata/conf.d`).

   the static installer now removes files in `etc/netdata` that match the stock MD5 hashes. It does this **only once** (so it remembers it has done it in the past and does not do it again).

   the static installer create the same links in `/opt/netdata/etc/netdata` as the git one.

- [x] adapt `netdata.spec.in` by @paulfantom

### CODE REVIEW

**Comment by paulfantom** [comment]
Looking at a number of modules written in C which need the same mechanism of getting config, I wonder if we could have a library file (like `config.[c|h]`) and just reuse it in those modules?

**Comment by ktsaou** [comment]
> Looking at a number of modules written in C which need the same mechanism of getting config, I wonder if we could have a library file (like `config.[c|h]`) and just reuse it in those modules?

All C modules are using netdata.conf.

Only health config and statsd config is somewhat complex. And I am trying to come up with a common solution for both (they both read all the files in all subdirectories of a path).

**Comment by paulfantom** [comment]
That's 2-3 places where we will have the same system for getting configuration implemented, for me that's enough to move it into a library. But final decision is yours.

**Comment by ktsaou** [comment]
> That's 2-3 places where we will have the same system for getting configuration implemented, for me that's enough to move it into a library. But final decision is yours.

exactly. I am trying to create a common library for both. But not link it to a `.so` or `.a`.
Most of netdata core is library code, though not separated from the project. Later maybe...

**Comment by ktsaou** [comment]
The problem with this is that it will have to be merged at once.
I can do all the tasks above, except the python one...

**Comment by paulfantom** [comment]
> The problem with this is that it will have to be merged at once.

Why not merge it into a `devel` branch and after development on this is done merge `devel` into `master`? It will even have a benefit of providing smaller, more atomic PRs which are easier to review.

**Comment by ktsaou** [comment]
> Why not merge it into a `devel` branch and after development on this is done merge `devel` into `master`? It will even have a benefit of providing smaller, more atomic PRs which are easier to review.

I like the idea of a `devel` branch, but I thought the whole idea was that even `devel` has always to be stable, meaning that we should not merge something we know is broken. And if we merge the above without the python part, it will be broken.

Can 2+ devs contribute to the same PR?

**Comment by paulfantom** [comment]
> Can 2+ devs contribute to the same PR?

Yes they can. Second developer can clone your fork and commit to this branch. But this still will create a very large PR at least in case of this PR.

> I like the idea of a devel branch, but I thought the whole idea was that even devel has always to be stable

"Devel" branch can be unstable and broken, that's why it is called "devel". "Master" on the other hand should be always stable. There is no reason of duplicating this behavior by having two branches which are used for the same purpose.

BTW, using approach of merging everything to "devel" and then this branch to "master" is not a recommended workflow in GitHub, but it is something that is called git-flow. More on this subject is in GitLab tutorial: https://about.gitlab.com/2014/09/29/gitlab-flow/

**Comment by ktsaou** [comment]
Yes, I know about the `develop` branch in gitflow. But still I think it should only get complete features. Check this: https://datasift.github.io/gitflow/IntroducingGitFlow.html

Incomplete features should go to feature branches. Develop should be production ready at all times.

At least this is how I understand gitflow...

**Comment by paulfantom** [comment]
Nonetheless git-flow just adds unnecessary problems.

**Comment by paulfantom** [comment]
And just do `s/devel/feature_branch/` in my previous comments and you will have what I meant in git-flow conventions.

**Comment by ktsaou** [comment]
@alonbl I need your help on this:

This PR changes netdata so that its stock config files are now installed in `/usr/lib/netdata/conf.d`, instead of `/etc/netdata`. We have added code so that netdata automatically loads the stock config file from `/usr/lib/netdata/conf.d` when the file required is not found under `/etc/netdata`.

As you understand this fixes a long standing issue we had. I believe this was your suggestion.

Anyway, I don't know how to modify `netdata.spec.in` for this.

I remind you that `make install` installs the stock config files (now in `/usr/lib/netdata/conf.d/`).

So, what needs to be changed in `netdata.spec.in` to pick up the files at the new position and set proper permissions?

We also need `netdata.spec.in` to create empty directories under `/etc/netdata` so that people will have a clue what to copy and where.

Thank you!

**Comment by ktsaou** [comment]
This is almost done.

If we fix `python.d.plugin`, then updating to this version should not introduce any change, apart from `/etc/netdata` becoming suddenly empty of stock configs.

Until we fix `python.d.plugin`, installing this PR will leave python plugin unconfigured.

So, @l2isbad I need your help, to fix `python.d.plugin` and start testing it end to end...

**Comment by paulfantom** [comment]
Probably this needs to be done:

```diff
diff --git a/netdata.spec.in b/netdata.spec.in
index 36a8530e..373464ab 100644
--- a/netdata.spec.in
+++ b/netdata.spec.in
@@ -181,17 +181,18 @@ rm -rf "${RPM_BUILD_ROOT}"
 %defattr(-,root,root)
 
 %dir %{_sysconfdir}/%{name}
-
-%config(noreplace) %{_sysconfdir}/%{name}/*.conf
-%config(noreplace) %{_sysconfdir}/%{name}/charts.d/*.conf
-%config(noreplace) %{_sysconfdir}/%{name}/health.d/*.conf
-#%%config(noreplace) %{_sysconfdir}/%{name}/node.d/*.conf
-%config(noreplace) %{_sysconfdir}/%{name}/python.d/*.conf
-%config(noreplace) %{_sysconfdir}/%{name}/statsd.d/*.conf
+%dir %{_libdir}/%{name}
+
+%config(noreplace) %{_libdir}/%{name}/*.conf
+%config(noreplace) %{_libdir}/%{name}/charts.d/*.conf
+%config(noreplace) %{_libdir}/%{name}/health.d/*.conf
+#%%config(noreplace) %{_libdir}/%{name}/node.d/*.conf
+%config(noreplace) %{_libdir}/%{name}/python.d/*.conf
+%config(noreplace) %{_libdir}/%{name}/statsd.d/*.conf
 %config(noreplace) %{_sysconfdir}/logrotate.d/%{name}
 
 # To be eventually moved to %%_defaultdocdir
-%{_sysconfdir}/%{name}/node.d/*.md
+%{_libdir}/%{name}/node.d/*.md
 %{_libexecdir}/%{name}
 %{_sbindir}/%{name}
```

**Comment by l2isbad** [comment]
@ktsaou done

**Comment by ktsaou** [comment]
@l2isbad 2 changes please:

1. When it fails to load a file, please make it log an error, showing the filename that failed to be found / loaded. So, if there is no user config for something, an error should always be shown, **with the filename**.

2. In debug mode it should show the filenames it loads.

**Comment by ktsaou** [comment]
@paulfantom push it please. This is a `feature_branch` (though on my fork - I don't know how to merge on github on another branch anyway...).

You will need a clone of this, since it needs a lot of testing and I am looking forward for your help...

**Comment by ktsaou** [comment]
@l2isbad thanks.

Please check this:

```sh
# /usr/libexec/netdata/plugins.d/python.d.plugin 1
2018-09-26 13:47:12: python.d INFO: plugin: main: Using python 3
2018-09-26 13:47:12: python.d ERROR: plugin: main: failed to load '/etc/netdata/python.d/apache.conf'
2018-09-26 13:47:12: python.d ERROR: plugin: main: failed to load '/etc/netdata/python.d/beanstalk.conf'
2018-09-26 13:47:12: python.d ERROR: plugin: main: failed to load '/etc/netdata/python.d/bind_rndc.conf'
2018-09-26 13:47:12: python.d ERROR: plugin: main: failed to load '/etc/netdata/python.d/boinc.conf'
2018-09-26 13:47:12: python.d ERROR: plugin: main: failed to load '/etc/netdata/python.d/ceph.conf'
2018-09-26 13:47:12: python.d ERROR: plugin: main: failed to load '/etc/netdata/python.d/couchdb.conf'
2018-09-26 13:47:12: python.d ERROR: plugin: main: failed to load '/etc/netdata/python.d/cpufreq.conf'
2018-09-26 13:47:12: python.d ERROR: plugin: main: failed to load '/etc/netdata/python.d/cpuidle.conf'
2018-09-26 13:47:12: python.d ERROR: plugin: main: failed to load '/usr/lib/netdata/conf.d/python.d/cpuidle.conf': FileNotFoundError(2, 'No such file or directory')
2018-09-26 13:47:12: python.d ERROR: plugin: main: module load config: 'cpuidle' => [FAILED]
2018-09-26 13:47:12: python.d ERROR: plugin: main: failed to load '/etc/netdata/python.d/dns_query_time.conf'
2018-09-26 13:47:12: python.d ERROR: plugin: main: failed to load '/etc/netdata/python.d/dnsdist.conf'
2018-09-26 13:47:12: python.d ERROR: plugin: main: failed to load '/etc/netdata/python.d/dockerd.conf'
2018-09-26 13:47:12: python.d ERROR: plugin: main: failed to load '/etc/netdata/python.d/dovecot.conf'
2018-09-26 13:47:12: python.d ERROR: plugin: main: failed to load '/etc/netdata/python.d/elasticsearch.conf'
2018-09-26 13:47:12: python.d ERROR: plugin: main: failed to load '/etc/netdata/python.d/exim.conf'
2018-09-26 13:47:12: python.d ERROR: plugin: main: failed to load '/etc/netdata/python.d/fail2ban.conf'
2018-09-26 13:47:12: python.d ERROR: plugin: main: failed to load '/etc/netdata/python.d/freeradius.conf'
2018-09-26 13:47:12: python.d ERROR: plugin: main: failed to load '/etc/netdata/python.d/haproxy.conf'
2018-09-26 13:47:12: python.d ERROR: plugin: main: failed to load '/etc/netdata/python.d/hddtemp.conf'
2018-09-26 13:47:12: python.d ERROR: plugin: main: failed to load '/etc/netdata/python.d/httpcheck.conf'
```

There is no reason on this:

```
2018-09-26 13:47:12: python.d ERROR: plugin: main: failed to load '/etc/netdata/python.d/apache.conf'
```

There are 2 logs for this:

```
2018-09-26 13:47:12: python.d ERROR: plugin: main: failed to load '/usr/lib/netdata/conf.d/python.d/cpuidle.conf': FileNotFoundError(2, 'No such file or directory')
2018-09-26 13:47:12: python.d ERROR: plugin: main: module load config: 'cpuidle' => [FAILED]
```

Please only 1 log, with reason.

**Comment by ktsaou** [comment]
1 log with **reason and filename**

**Comment by ktsaou** [comment]
And please change `failed to load ...` with `cannot load ...`
If you have separate log code for the 2 files, please also append `Will try stock version` on the failure log of user-configs.

**Comment by ktsaou** [comment]
The whole idea is to help people find their way on this. It is quite confusing already so, we need better error messages to help them.

**Comment by paulfantom** [comment]
> ERROR: plugin: main: failed to load '/etc/netdata/python.d/apache.conf'

This is not an ERROR if we have stock config in `/lib/...`

**Comment by l2isbad** [comment]
Indeed. `INFO` then?

**Comment by ktsaou** [comment]
> This is not an ERROR if we have stock config in `/lib/...`

Agree, if we can load the stock config. But this error is printed before loading the stock config, so we can't tell. And the fact that there is a file at the place we expect the stock config, does not mean we will be able to load it.

It makes it quite complex. So, I followed a simple rule. Any config file that we try to load, that cannot be loaded for any reason, gives an ERROR.

In shell scripts we have also warnings, so I used warnings for "file not found" and errors when the file is found but cannot be read.
In C I used errors.
In node I used errors.

**Comment by Ferroin** [comment]
Why not:

* If a config file can't be loaded because it's not there, it's a warning.
* If it can't be loaded because of invalid syntax, it's an error.
* If it was successfully loaded, it's a debug message.
* After all locations have been searched, there's either an info message indicating which one is being used, or an error indicating that it couldn't find a valid config.

That way the user knows what config file is being used (or if none was found), and gets some info on what config files weren't usable.

**Comment by ktsaou** [comment]
@Ferroin I agree. But we don't provide `warning` in the code. Either `INFO` or `ERROR`. We need to extend a bit the logging to support warnings too.

**Comment by ktsaou** [comment]
I believe this is ready.
I have tested everything, but please also make a few tests.

**Comment by ktsaou** [comment]
Only `netdata.spec` is missing.

**Comment by Ferroin** [comment]
From a cursory glance, this looks fine to me.  I'll pull down a copy of it and run some testing at some point later today on one of my systems, and have some automated tests check it on a couple of different distros overnight and get back with you first thing tomorrow with the results.

**Comment by paulfantom** [comment]
`git apply` done :)

**Comment by ktsaou** [comment]
@Ferroin perfect!
The more testing the better.
Also, please try vanilla installations (not updates).

File: netdata.spec.in [hunk 1]
```diff
@@ -181,17 +181,18 @@ rm -rf "${RPM_BUILD_ROOT}"
...
-%config(noreplace) %{_sysconfdir}/%{name}/python.d/*.conf
-%config(noreplace) %{_sysconfdir}/%{name}/statsd.d/*.conf
+%dir %{_libdir}/%{name}
+
+%config(noreplace) %{_libdir}/%{name}/*.conf
+%config(noreplace) %{_libdir}/%{name}/charts.d/*.conf
+%config(noreplace) %{_libdir}/%{name}/health.d/*.conf
+#%%config(noreplace) %{_libdir}/%{name}/node.d/*.conf
+%config(noreplace) %{_libdir}/%{name}/python.d/*.conf
+%config(noreplace) %{_libdir}/%{name}/statsd.d/*.conf
```
Comment (ktsaou) [inline netdata.spec.in]: @paulfantom I think `noreplace` will do what it says, but we need it to replace these files.

File: netdata.spec.in [hunk 2]
```diff
@@ -215,9 +216,13 @@ rm -rf "${RPM_BUILD_ROOT}"
 
 %dir %{_datadir}/%{name}
 %dir %{_sysconfdir}/%{name}/health.d
+%dir %{_libdir}/%{name}/health.d
 %dir %{_sysconfdir}/%{name}/python.d
+%dir %{_libdir}/%{name}/python.d
 %dir %{_sysconfdir}/%{name}/charts.d
+%dir %{_libdir}/%{name}/charts.d
 %dir %{_sysconfdir}/%{name}/node.d
+%dir %{_libdir}/%{name}/node.d
```
Comment (ktsaou) [inline netdata.spec.in]: it seems `statsd.d` is missing.

File: netdata.spec.in (hunk 1 above)
Comment (paulfantom) [reply netdata.spec.in]: I don't know, maybe @alonbl can help?

**Comment by Ferroin** [comment]
@ktsaou Already part of the plan.  Normally when I set things up to check installation on my various distro VM's, I have it run three checks on each VM:

* Clean install of the master branch built from sources works (if this fails, I know the others should fail too and just skip them).
* Clean install of the branch being tested built from sources works.
* Update to a clean install of the master branch using the branch being tested, both built from sources, works.

**Comment by paulfantom** [comment]
@Ferroin First point is done by CI every day, second is done on every PR, third should be probably automated somewhere in the future.

**Comment by paulfantom** [comment]
You can check what CI is running against by going for example to https://travis-ci.com/netdata/netdata/builds/86020763

File: netdata.spec.in [hunk 3]
```diff
@@ -223,6 +223,9 @@ rm -rf "${RPM_BUILD_ROOT}"
 %dir %{_libdir}/%{name}/charts.d
 %dir %{_sysconfdir}/%{name}/node.d
 %dir %{_libdir}/%{name}/node.d
+%dir %{_sysconfdir}/%{name}/stats.d
+%dir %{_libdir}/%{name}/stats.d
```
Comment (ktsaou) [inline netdata.spec.in]: oops! This should be `statsd.d` (a `d` is missing).

**Comment by ktsaou** [comment]
merged!

**Pull request merged**